src/Components/Web/src/Forms/EditContextModelBindingContextExtensions.cs

```diff
@@ -0,0 +1,102 @@
+    {
+        ArgumentNullException.ThrowIfNull(context, nameof(context));
+        ArgumentNullException.ThrowIfNull(bindingContext, nameof(bindingContext));
+
+        context.Properties[_key] = bindingContext;
+
+        return new BindingContextEventSubscriptions(context, bindingContext);
+    }
+
+    /// <summary>
+    /// Gets the attempted value for the specified field name.
+    /// </summary>
+    /// <param name="context">The <see cref="EditContext"/>.</param>
+    /// <param name="fieldName">The field name.</param>
+    /// <returns></returns>
+    public static string? GetAttemptedValue(this EditContext context, string fieldName)
+    {
+        ArgumentNullException.ThrowIfNull(context, nameof(context));
+        ArgumentNullException.ThrowIfNull(fieldName, nameof(fieldName));
+
+        if (context.Properties.TryGetValue(_key, out var result) && result is ModelBindingContext bindingContext)
+        {
+            return bindingContext.GetAttemptedValue(fieldName);
+        }
+
+        return null;
+    }
+
+    private sealed class BindingContextEventSubscriptions : IDisposable
+    {
+        private readonly EditContext _editContext;
+        private readonly ModelBindingContext _bindingContext;
+        private ValidationMessageStore? _messages;
+
+        public BindingContextEventSubscriptions(EditContext editContext, ModelBindingContext serviceProvider)
+        {
+            _editContext = editContext;
+            _bindingContext = serviceProvider;
+
+            _editContext.OnValidationRequested += OnValidationRequested;
+        }
+
+        private void OnValidationRequested(object? sender, ValidationRequestedEventArgs e)
+        {
+            if (_messages != null)
+            {
+                // We already added the messages from the binding context,
+                // we don't have to do anything.
+                return;
+            }
+
+            _messages = new ValidationMessageStore(_editContext);
+            var adddedMessages = false;
+            foreach (var ((owner, key), errors) in _bindingContext.GetAllErrors())
+            {
+                FieldIdentifier fieldIdentifier;
+                fieldIdentifier = new FieldIdentifier(owner ?? _editContext.Model, key);
+
+                foreach (var error in errors)
+                {
+                    adddedMessages = true;
+                    // TODO: We need to support localizing the error message.
+                    _messages.Add(fieldIdentifier, error.ToString(CultureInfo.CurrentCulture));
+                }
+            }
+
+            if (adddedMessages)
+            {
+                // There were binding errors, notify.
+                _editContext.NotifyValidationStateChanged();
+            }
+        }
+
+        public void Dispose()
+        {
+            _messages?.Clear();
+            _editContext.OnValidationRequested -= OnValidationRequested;
+            _editContext.NotifyValidationStateChanged();
```

Comment:
I think you could call this conditionally based on whether `_messages` was empty.

Reply:
done